File: management-domain/Sales-department/Backend/Node.js/sales-automation-service/src/shared/context/request-context.ts
import { createParamDecorator, ExecutionContext } from '@nestjs/common';

/**
 * Request context for multi-tenancy support
 * Contains tenant, user, and correlation information
 */
export interface RequestContextData {
  tenantId: string;
  userId: string;
  correlationId: string;
  roles: string[];
  organizationId?: string;
}

/**
 * Custom decorator to extract request context from request
 */
export const RequestContext = createParamDecorator(
  (data: unknown, ctx: ExecutionContext): RequestContextData => {
    const request = ctx.switchToHttp().getRequest();
    return {
      tenantId: request.headers['x-tenant-id'] || request.tenantId || 'default',
      userId: request.headers['x-user-id'] || request.userId || 'system',
      correlationId: request.headers['x-correlation-id'] || request.correlationId || generateCorrelationId(),
      roles: request.headers['x-roles']?.split(',') || request.roles || [],
      organizationId: request.headers['x-organization-id'] || request.organizationId,
    };
  },
);

function generateCorrelationId(): string {
  return `corr-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
}

/**
 * Request context holder class for storing context in async local storage
 */
import { AsyncLocalStorage } from 'async_hooks';

export const RequestContextStorage = new AsyncLocalStorage<RequestContextData>();

export class RequestContextHolder {
  static setContext(context: RequestContextData): void {
    RequestContextStorage.enterWith(context);
  }

  static getContext(): RequestContextData | undefined {
    return RequestContextStorage.getStore();
  }

  static getTenantId(): string {
    return RequestContextStorage.getStore()?.tenantId || 'default';
  }

  static getUserId(): string {
    return RequestContextStorage.getStore()?.userId || 'system';
  }

  static getCorrelationId(): string {
    return RequestContextStorage.getStore()?.correlationId || generateCorrelationId();
  }

  static getOrganizationId(): string | undefined {
    return RequestContextStorage.getStore()?.organizationId;
  }

  static hasRole(role: string): boolean {
    const roles = RequestContextStorage.getStore()?.roles || [];
    return roles.includes(role);
  }
}

/**
 * Execution context interface for use cases and services
 */
export interface ExecutionContext {
  tenantId: string;
  userId: string;
  correlationId: string;
  roles?: string[];
  organizationId?: string;
}

export class ExecutionContextImpl implements ExecutionContext {
  constructor(
    public tenantId: string = 'default',
    public userId: string = 'system',
    public correlationId: string = generateCorrelationId(),
    public roles?: string[],
    public organizationId?: string,
  ) {}
}

File: shared-business-logics/services/universal-tracking-services/Backend/Java/universal-tracking-service/src/main/java/com/gogidix/shared/security/context/RequestContext.java
package com.gogidix.shared.security.context;

import java.util.Optional;
import java.util.UUID;

/**
 * Request context holder for tenant and user information.
 */
public class RequestContext {

    private static final ThreadLocal<String> TENANT_ID = new ThreadLocal<>();
    private static final ThreadLocal<String> USER_ID = new ThreadLocal<>();
    private static final ThreadLocal<String> CORRELATION_ID = new ThreadLocal<>();

    /**
     * Set tenant ID for current request
     */
    public static void setTenantId(String tenantId) {
        TENANT_ID.set(tenantId);
    }

    /**
     * Get tenant ID for current request
     */
    public static Optional<String> getTenantId() {
        return Optional.ofNullable(TENANT_ID.get());
    }

    /**
     * Set user ID for current request
     */
    public static void setUserId(String userId) {
        USER_ID.set(userId);
    }

    /**
     * Get user ID for current request
     */
    public static Optional<String> getUserId() {
        return Optional.ofNullable(USER_ID.get());
    }

    /**
     * Set correlation ID for current request
     */
    public static void setCorrelationId(String correlationId) {
        CORRELATION_ID.set(correlationId);
    }

    /**
     * Get correlation ID for current request
     */
    public static Optional<String> getCorrelationId() {
        return Optional.ofNullable(CORRELATION_ID.get());
    }

    /**
     * Generate and set a new correlation ID
     */
    public static String generateCorrelationId() {
        String correlationId = UUID.randomUUID().toString();
        setCorrelationId(correlationId);
        return correlationId;
    }

    /**
     * Clear all context data
     */
    public static void clear() {
        TENANT_ID.remove();
        USER_ID.remove();
        CORRELATION_ID.remove();
    }
}
